### PR TITLE
MINOR: Add select changes from 3rd KIP-307 PR for incrementing name index counter

### PR DESCRIPTION
**Pull request by bbejeck**
When users provide a name for operation via the Streams DSL, we need to increment the counter used for auto-generated names to make sure any operators downstream of a named operator still produce a compatible name.  

This PR is a subset of #6411 by @fhussonnois.  We need to merge this PR now because it covers cases when users name repartition topics or state stores.  

Updated tests to reflect the counter produces expected number even when the user provides a name.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
ping @guozhangwang @mjsax @vvcephei @ableegoldman @cadonna @abbccdda  for review

\cc @fhussonnois

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 1]
```diff
@@ -440,8 +440,8 @@ public void shouldUseSpecifiedNameForTableSourceProcessor() {
                 topology,
                 expected,
                 expected + "-table-source",
-                "KSTREAM-SOURCE-0000000002",
-                "KTABLE-SOURCE-0000000003");
+                "KSTREAM-SOURCE-0000000004",
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:443]: These number changes are valid as 2 store names (internal, not visible for IQ) are generated.  I validated the topology description against a description generated by `2.2`

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 2]
```diff
@@ -424,7 +424,7 @@ public void shouldUseSpecifiedNameForStreamSourceProcessor() {
         builder.stream(STREAM_TOPIC_TWO);
         builder.build();
         final ProcessorTopology topology = builder.internalTopologyBuilder.rewriteTopology(new StreamsConfig(props)).build();
-        assertSpecifiedNameForOperation(topology, expected, "KSTREAM-SOURCE-0000000000");
+        assertSpecifiedNameForOperation(topology, expected, "KSTREAM-SOURCE-0000000001");
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:427]: First source node would be `KSTREAM-SOURCE-0000000000` so this is a valid change in the test

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 3]
```diff
@@ -467,7 +467,7 @@ public void shouldUseSpecifiedNameForSinkProcessor() {
         stream.to(STREAM_TOPIC_TWO);
         builder.build();
         final ProcessorTopology topology = builder.internalTopologyBuilder.rewriteTopology(new StreamsConfig(props)).build();
-        assertSpecifiedNameForOperation(topology, "KSTREAM-SOURCE-0000000000", expected, "KSTREAM-SINK-0000000001");
+        assertSpecifiedNameForOperation(topology, "KSTREAM-SOURCE-0000000000", expected, "KSTREAM-SINK-0000000002");
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:470]: The first source is `KSTREAM-SOURCE-0000000000` second would be `...01` so this change is correct.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:443]: I see. So the diff we're seeing is against tests that were previously modified (incorrectly) to match the non-incrementing behavior that you're fixing right now?

Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:443]: Yes, that is correct.

**Comment by vvcephei** [comment]
Hey @bbejeck , thanks for fixing this!

I found the logic a little hard to follow, so I pulled it down to play around with it. I think that the original decision to use the Optional FP logic instead of branches, and also making the provider nullable (which was only in support of one test :/ ), was what was making it so convoluted.

What do you think about this: https://github.com/bbejeck/kafka/pull/1 ?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:443]: perfect. thanks for the reassurance!

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:443]: Actually, those test are newly added in a previous KIP-307 PR... But if Bill check agains 2.2, it should be fine.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by mjsax** [comment]
Still LGTM

**Comment by bbejeck** [comment]
Merged #6754 into trunk.
